src/RenderAPI/EspSwapChain.cc

```diff
@@ -0,0 +1,489 @@
+
+			VkImageViewCreateInfo view_info{};
+			view_info.sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO;
+			view_info.image = m_depth_images[i];
+			view_info.viewType = VK_IMAGE_VIEW_TYPE_2D;
+			view_info.format = depth_format;
+			view_info.subresourceRange.aspectMask = VK_IMAGE_ASPECT_DEPTH_BIT;
+			view_info.subresourceRange.baseMipLevel = 0;
+			view_info.subresourceRange.levelCount = 1;
+			view_info.subresourceRange.baseArrayLayer = 0;
+			view_info.subresourceRange.layerCount = 1;
+
+			if (vkCreateImageView(m_device.get_device(), &view_info, nullptr, &m_depth_image_views[i]) != VK_SUCCESS)
+			{
+				ESP_CORE_ERROR("Failed to create texture image view");
+				throw std::runtime_error("Failed to create texture image view");
+			}
+		}
+	}
+
+	void EspSwapChain::create_sync_objects()
+	{
+		m_image_available_semaphores.resize(MAX_FRAMES_IN_FLIGHT);
+		m_render_finished_semaphores.resize(MAX_FRAMES_IN_FLIGHT);
+		m_in_flight_fences.resize(MAX_FRAMES_IN_FLIGHT);
+		m_images_in_flight.resize(image_count(), VK_NULL_HANDLE);
+
+		VkSemaphoreCreateInfo semaphore_info = {};
+		semaphore_info.sType = VK_STRUCTURE_TYPE_SEMAPHORE_CREATE_INFO;
+
+		VkFenceCreateInfo fence_info = {};
+		fence_info.sType = VK_STRUCTURE_TYPE_FENCE_CREATE_INFO;
+		fence_info.flags = VK_FENCE_CREATE_SIGNALED_BIT;
+
+		for (size_t i = 0; i < MAX_FRAMES_IN_FLIGHT; i++)
+		{
+			if (vkCreateSemaphore(m_device.get_device(), &semaphore_info, nullptr, &m_image_available_semaphores[i]) != VK_SUCCESS
+				|| vkCreateSemaphore(m_device.get_device(), &semaphore_info, nullptr, &m_render_finished_semaphores[i]) != VK_SUCCESS
+				|| vkCreateFence(m_device.get_device(), &fence_info, nullptr, &m_in_flight_fences[i]) != VK_SUCCESS)
+			{
+				ESP_CORE_ERROR("Failed to create synchronization objects for a frame");
+				throw std::runtime_error("Failed to create synchronization objects for a frame");
+			}
+		}
+	}
+
+	VkSurfaceFormatKHR EspSwapChain::choose_swap_chain_surface_format(
+		const std::vector<VkSurfaceFormatKHR>& available_formats)
+	{
+		for (const auto& available_format : available_formats)
+		{
+			if (available_format.format == VK_FORMAT_B8G8R8A8_SRGB &&
+				available_format.colorSpace == VK_COLOR_SPACE_SRGB_NONLINEAR_KHR)
+			{
+				return available_format;
+			}
+		}
+
+		return available_formats[0];
+	}
+
+	VkPresentModeKHR EspSwapChain::choose_swap_chain_present_mode(
+		const std::vector<VkPresentModeKHR>& available_present_modes)
+	{
+		for (const auto& available_present_mode : available_present_modes)
+		{
+			if (available_present_mode == VK_PRESENT_MODE_MAILBOX_KHR)
+			{
+				log_chosen_swap_chain_present_mode("Mailbox");
+				return available_present_mode;
+			}
+		}
+
+		// for (const auto &available_present_mode : available_present_modes) {
+		//   if (available_present_mode == VK_PRESENT_MODE_IMMEDIATE_KHR)
+		//   {
+		//		log_chosen_swap_chain_present_mode("Immediate");
+		//   	return available_present_mode;
+		//   }
+		// }
```

Comment:
Why is this part commented out?

Reply:
Immediate mode allows to swap frame any time. It allows GPU to work at full capacity and I think it is used for example in VR, but I guess there is no benefits from it in our case so we can remove it.